File: apps/agor-android/app/src/main/kotlin/live/agor/app/models/Session.kt
package live.agor.app.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class SessionStatus {
    @SerialName("idle") IDLE,
    @SerialName("running") RUNNING,
    @SerialName("stopping") STOPPING,
    @SerialName("awaiting_permission") AWAITING_PERMISSION,
    @SerialName("awaiting_input") AWAITING_INPUT,
    @SerialName("timed_out") TIMED_OUT,
    @SerialName("completed") COMPLETED,
    @SerialName("failed") FAILED;

    val needsAttention: Boolean
        get() = this == AWAITING_PERMISSION || this == AWAITING_INPUT

    val isActive: Boolean
        get() = this == RUNNING || this == STOPPING || needsAttention

    val displayLabel: String
        get() = when (this) {
            IDLE -> "Idle"
            RUNNING -> "Running"
            STOPPING -> "Stopping"
            AWAITING_PERMISSION -> "Awaiting Permission"
            AWAITING_INPUT -> "Awaiting Input"
            TIMED_OUT -> "Timed Out"
            COMPLETED -> "Completed"
            FAILED -> "Failed"
        }
}

@Serializable
enum class AgenticTool {
    @SerialName("claude-code") CLAUDE_CODE,
    @SerialName("codex") CODEX,
    @SerialName("gemini") GEMINI,
    @SerialName("opencode") OPENCODE;

    val displayName: String
        get() = when (this) {
            CLAUDE_CODE -> "Claude Code"
            CODEX -> "Codex"
            GEMINI -> "Gemini"
            OPENCODE -> "OpenCode"
        }
}

@Serializable
enum class PermissionMode {
    @SerialName("default") DEFAULT,
    @SerialName("acceptEdits") ACCEPT_EDITS,
    @SerialName("bypassPermissions") BYPASS,
    @SerialName("plan") PLAN,
    @SerialName("dontAsk") DONT_ASK,
    @SerialName("autoEdit") AUTO_EDIT,
    @SerialName("yolo") YOLO,
    @SerialName("ask") ASK,
    @SerialName("auto") AUTO,
    @SerialName("on-failure") ON_FAILURE,
    @SerialName("allow-all") ALLOW_ALL,
}

@Serializable
data class GitState(
    val ref: String? = null,
    @SerialName("base_sha") val baseSha: String? = null,
    @SerialName("current_sha") val currentSha: String? = null,
)

@Serializable
data class SessionGenealogy(
    @SerialName("forked_from_session_id") val forkedFromSessionId: String? = null,
    @SerialName("fork_point_task_id") val forkPointTaskId: String? = null,
    @SerialName("fork_point_message_index") val forkPointMessageIndex: Int? = null,
    @SerialName("parent_session_id") val parentSessionId: String? = null,
    @SerialName("spawn_point_task_id") val spawnPointTaskId: String? = null,
    @SerialName("spawn_point_message_index") val spawnPointMessageIndex: Int? = null,
    val children: List<String> = emptyList(),
)

@Serializable
data class PermissionConfig(
    val mode: PermissionMode? = null,
)

@Serializable
data class ModelConfig(
    val mode: String? = null,
    val model: String? = null,
    val provider: String? = null,
    val notes: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null,
    val thinkingMode: String? = null,
    val manualThinkingTokens: Int? = null,
    val effort: String? = null,
)

@Serializable
data class Session(
    @SerialName("session_id") val sessionId: String,
    @SerialName("agentic_tool") val agenticTool: AgenticTool,
    @SerialName("agentic_tool_version") val agenticToolVersion: String? = null,
    @SerialName("sdk_session_id") val sdkSessionId: String? = null,
    val status: SessionStatus,
    @SerialName("created_at") val createdAt: String,
    @SerialName("last_updated") val lastUpdated: String,
    @SerialName("created_by") val createdBy: String,
    @SerialName("unix_username") val unixUsername: String? = null,
    @SerialName("worktree_id") val worktreeId: String,
    @SerialName("worktree_board_id") val worktreeBoardId: String? = null,
    val url: String? = null,
    @SerialName("git_state") val gitState: GitState? = null,
    val genealogy: SessionGenealogy? = null,
    val tasks: List<String>? = null,
    @SerialName("message_count") val messageCount: Int? = null,
    val title: String? = null,
    val description: String? = null,
    @SerialName("permission_config") val permissionConfig: PermissionConfig? = null,
    @SerialName("model_config") val modelConfig: ModelConfig? = null,
    @SerialName("current_context_usage") val currentContextUsage: Int? = null,
    @SerialName("context_window_limit") val contextWindowLimit: Int? = null,
    @SerialName("scheduled_from_worktree") val scheduledFromWorktree: Boolean? = null,
    @SerialName("ready_for_prompt") val readyForPrompt: Boolean? = null,
    val archived: Boolean? = null,
    @SerialName("archived_reason") val archivedReason: String? = null,
) {
    val displayTitle: String
        get() {
            title?.takeIf { it.isNotEmpty() }?.let { return it }
            description?.takeIf { it.isNotEmpty() }?.let { return it }
            return "Session ${sessionId.take(8)}"
        }

    val hasExplicitTitle: Boolean
        get() {
            val t = title ?: return false
            if (t.isEmpty()) return false
            if (t.startsWith("[Scheduled run")) return false
            return true
        }

    val isPlanMode: Boolean
        get() = permissionConfig?.mode == PermissionMode.PLAN

    val isPromptable: Boolean
        get() = status == SessionStatus.IDLE || readyForPrompt == true

    val isScheduled: Boolean
        get() = scheduledFromWorktree == true || title?.startsWith("[Scheduled ") == true
}

File: apps/agor-rust-android/crates/agor/src/models/session.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SessionStatus {
    Idle,
    Running,
    Stopping,
    AwaitingPermission,
    AwaitingInput,
    TimedOut,
    Completed,
    Failed,
}

impl SessionStatus {
    pub fn needs_attention(&self) -> bool {
        matches!(self, Self::AwaitingPermission | Self::AwaitingInput)
    }

    pub fn is_active(&self) -> bool {
        matches!(self, Self::Running | Self::Stopping | Self::AwaitingPermission | Self::AwaitingInput)
    }

    pub fn display_label(&self) -> &str {
        match self {
            Self::Idle => "Idle",
            Self::Running => "Running",
            Self::Stopping => "Stopping",
            Self::AwaitingPermission => "Needs Permission",
            Self::AwaitingInput => "Needs Input",
            Self::TimedOut => "Timed Out",
            Self::Completed => "Completed",
            Self::Failed => "Failed",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum AgenticTool {
    ClaudeCode,
    Codex,
    Gemini,
    Opencode,
}

impl AgenticTool {
    pub fn display_name(&self) -> &str {
        match self {
            Self::ClaudeCode => "Claude Code",
            Self::Codex => "Codex",
            Self::Gemini => "Gemini",
            Self::Opencode => "OpenCode",
        }
    }

    pub fn permission_modes(&self) -> Vec<PermissionMode> {
        match self {
            Self::ClaudeCode => vec![
                PermissionMode::Default,
                PermissionMode::AcceptEdits,
                PermissionMode::Bypass,
                PermissionMode::Plan,
            ],
            Self::Codex => vec![
                PermissionMode::Ask,
                PermissionMode::Auto,
                PermissionMode::OnFailure,
                PermissionMode::AllowAll,
            ],
            Self::Gemini => vec![
                PermissionMode::Default,
                PermissionMode::AutoEdit,
                PermissionMode::Yolo,
            ],
            Self::Opencode => vec![PermissionMode::Default],
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PermissionMode {
    Default,
    AcceptEdits,
    Bypass,
    Plan,
    DontAsk,
    AutoEdit,
    Yolo,
    Ask,
    Auto,
    OnFailure,
    AllowAll,
}

impl PermissionMode {
    pub fn display_label(&self) -> &str {
        match self {
            Self::Default => "Default",
            Self::AcceptEdits => "Accept Edits",
            Self::Bypass => "Bypass Permissions",
            Self::Plan => "Plan Mode",
            Self::DontAsk => "Don't Ask",
            Self::AutoEdit => "Auto Edit",
            Self::Yolo => "YOLO",
            Self::Ask => "Ask",
            Self::Auto => "Auto",
            Self::OnFailure => "On Failure",
            Self::AllowAll => "Allow All",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GitState {
    #[serde(rename = "ref")]
    pub git_ref: Option<String>,
    pub base_sha: Option<String>,
    pub current_sha: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SessionGenealogy {
    pub forked_from_session_id: Option<String>,
    pub fork_point_task_id: Option<String>,
    pub fork_point_message_index: Option<i32>,
    pub parent_session_id: Option<String>,
    pub spawn_point_task_id: Option<String>,
    pub spawn_point_message_index: Option<i32>,
    #[serde(default)]
    pub children: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PermissionConfig {
    pub mode: Option<PermissionMode>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ModelConfig {
    pub mode: Option<String>,
    pub model: Option<String>,
    pub provider: Option<String>,
    pub notes: Option<String>,
    pub updated_at: Option<String>,
    pub thinking_mode: Option<String>,
    pub manual_thinking_tokens: Option<i32>,
    pub effort: Option<String>,
}

impl ModelConfig {
    pub fn display_summary(&self) -> String {
        let mut parts = Vec::new();
        if let Some(m) = &self.model {
            parts.push(m.clone());
        }
        if let Some(p) = &self.provider {
            parts.push(format!("via {p}"));
        }
        if let Some(e) = &self.effort {
            parts.push(format!("effort: {e}"));
        }
        if parts.is_empty() {
            "Default".to_string()
        } else {
            parts.join(" ")
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Session {
    pub session_id: String,
    pub agentic_tool: AgenticTool,
    pub agentic_tool_version: Option<String>,
    pub sdk_session_id: Option<String>,
    pub status: SessionStatus,
    pub created_at: String,
    pub last_updated: String,
    pub created_by: String,
    pub unix_username: Option<String>,
    pub worktree_id: String,
    pub worktree_board_id: Option<String>,
    pub url: Option<String>,
    pub git_state: Option<GitState>,
    pub genealogy: Option<SessionGenealogy>,
    pub tasks: Option<Vec<String>>,
    pub message_count: Option<i32>,
    pub title: Option<String>,
    pub description: Option<String>,
    pub permission_config: Option<PermissionConfig>,
    pub model_config: Option<ModelConfig>,
    pub current_context_usage: Option<i64>,
    pub context_window_limit: Option<i64>,
    pub scheduled_from_worktree: Option<bool>,
    pub ready_for_prompt: Option<bool>,
    pub archived: Option<bool>,
    pub archived_reason: Option<String>,
}

impl Session {
    pub fn display_title(&self) -> String {
        self.title
            .as_deref()
            .filter(|t| !t.is_empty())
            .unwrap_or("Untitled Session")
            .to_string()
    }

    pub fn has_explicit_title(&self) -> bool {
        self.title.as_deref().map_or(false, |t| !t.is_empty())
    }

    pub fn is_plan_mode(&self) -> bool {
        self.permission_config
            .as_ref()
            .and_then(|c| c.mode.as_ref())
            .map_or(false, |m| matches!(m, PermissionMode::Plan))
    }

    pub fn is_promptable(&self) -> bool {
        matches!(self.status, SessionStatus::Idle)
            || self.ready_for_prompt.unwrap_or(false)
    }

    pub fn can_queue_prompt(&self) -> bool {
        self.is_promptable() || self.status.is_active()
    }

    pub fn is_scheduled(&self) -> bool {
        self.scheduled_from_worktree.unwrap_or(false)
    }
}
